src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -2719,34 +2713,54 @@ namespace QgsWms
     }
   }
 
-  void QgsRenderer::setLayerFilter( QgsMapLayer *layer, const QStringList &filters ) const
+  void QgsRenderer::setLayerFilter( QgsMapLayer *layer, const QStringList &filters )
   {
     if ( layer->type() == QgsMapLayer::VectorLayer )
     {
       QgsVectorLayer *filteredLayer = qobject_cast<QgsVectorLayer *>( layer );
       Q_FOREACH ( QString filter, filters )
       {
-        if ( !testFilterStringSafety( filter ) )
+        if ( filter.startsWith( QStringLiteral( "<" ) ) && filter.endsWith( QStringLiteral( "Filter>" ) ) )
         {
-          throw QgsBadRequestException( QStringLiteral( "Filter string rejected" ),
-                                        QStringLiteral( "The filter string %1"
-                                            " has been rejected because of security reasons."
-                                            " Note: Text strings have to be enclosed in single or double quotes."
-                                            " A space between each word / special character is mandatory."
-                                            " Allowed Keywords and special characters are "
-                                            " AND,OR,IN,<,>=,>,>=,!=,',',(,),DMETAPHONE,SOUNDEX."
-                                            " Not allowed are semicolons in the filter expression." ).arg( filter ) );
+          // OGC filter
+          QDomDocument filterXml;
+          QString errorMsg;
+          if ( !filterXml.setContent( filter, true, &errorMsg ) )
+          {
+            throw QgsBadRequestException( QStringLiteral( "Filter string rejected" ),
+                                          QStringLiteral( "error message: %1. The XML string was: %2" ).arg( errorMsg, filter ) );
+          }
+          QDomElement filterElem = filterXml.firstChildElement();
+          QScopedPointer<QgsExpression> expression( QgsOgcUtils::expressionFromOgcFilter( filterElem ) );
+          mFeatureFilter.setFilter( filteredLayer, *expression );
+          filteredLayer->setSubsetString( "" );
```

Comment:
I think the line `filteredLayer->setSubsetString( "" );` has to be removed, because if the layer has a subsetstring, this line removed it and the result will not be the  combination of the OGC filter and the defined subsetstring.

Reply:
I don't agree. The `filteredLayer->setSubsetString(...)` is used only for the SQL style filters. But you cannot have both at the same time. And you don't want to inherit from the filter of the previous query if it was using a SQL style filter.

Reply:
the setSubsetstring will remove the one define in the project, it is not what the project creator want! Like in the case the filter is teh SQL one.

Reply:
OK, I didn't know about that. I've removed it. But where is it reset back to it's original value?